README
NAME
    CPAN::WWW::Testers::Generator - The pre-2009 CPAN Testers data manager codebase

DESCRIPTION

    This distribution used to generate the CPAN Testers database before 2009.

CPAN TESTERS

    The CPAN Testers project was conceived back in May 1998 by Graham Barr and 
    Chris Nandor as a way to provide multi-platform testing for modules uploaded
    to CPAN. Today there are over 20 million test reports and more than 100 
    testers regularly submitting reports each month, giving valuable feedback 
    for users and authors alike.
      
    The objective of the group is to test as many of the distributions 
    available on CPAN as possible, on as many platforms as possible, with a 
    variety of perl interpreters. The ultimate goal is to improve the 
    portability of the distributions on CPAN, and provide good feedback to the 
    authors.

    This distribution was originally written by Leon Brocard to download and
    summarize CPAN Testers data. However, since 2008, the codebase has been
    redesigned by Barbie, firstly to use the cpanstats database, and secondly
    to retrieve data from the Metabase API.

    Following the complete redesign of the tools in 2009, this distribution 
    and the tools used were no longer required.

    For the current set of releases for the CPAN Testers family of websites,
    please see:

    * CPAN-Testers-WWW-Reports
    * CPAN-Testers-WWW-Statistics
    * CPAN-Testers-WWW-Wiki
    * CPAN-Testers-WWW-Blog

    ADVANCED WARNING - ADVANCED WARNING - ADVANCED WARNING - ADVANCED WARNING

    This distribution will soon be deleted from CPAN. It is no longer
    applicable for the building of the CPAN Testers websites, and no longer
    maintained.

SEE ALSO

    * CPAN-Testers-WWW-Reports     - http://www.cpantesters.org
    * CPAN-Testers-WWW-Statistics  - http://stats.cpantesters.org
    * CPAN-Testers-WWW-Wiki        - http://wiki.cpantesters.org
    * CPAN-Testers-WWW-Blog        - http://blog.cpantesters.org

BECOME A TESTER
    The objective of the CPAN Testers is to test as many of the distributions 
    on CPAN as possible, on as many platforms as possible. The ultimate goal is
    to improve the portability of the distributions on CPAN, and provide good 
    feedback to the authors.

    Whether you have a common platform or a very unusual one, you can help by 
    testing modules you install and submitting reports. There are plenty of 
    module authors who could use test reports and helpful feedback on their 
    modules and distributions. 

    If you'd like to get involved, please take a look at the CPAN Testers Wiki,
    where you can learn how to install and configure one of the recommended 
    smoke tools.

    For further help and advice, please subscribe to the the CPAN Testers 
    discussion mailing list.

    CPAN Testers Wiki - http://wiki.cpantesters.org
    CPAN Testers Discuss mailing list
    - http://lists.cpan.org/showlist.cgi?name=cpan-testers-discuss

AUTHOR
    Original author:    Leon Brocard <acme@astray.com>   (C) 2002-2008
    Current maintainer: Barbie       <barbie@cpan.org>   (C) 2008-2012

LICENSE
    This code is distributed under the same license as Perl.


CHANGES
Revision history for Perl module CPAN::WWW::Testers::Generator.

0.31	29/04/2012
	- ADVANCED WARNING
	- - This distribution will be deleted shortly.
	- - It is no longer used to create the CPAN Testers website.
	- - Please see CPAN-Testers-Data-Generator.

0.30	01/10/2008
	- ensure upload pattern matching is for the end of the string.
	- add more option validation in bin/cpanstats-verify (v0.20).

0.29	19/09/2008
	- ignore any uploads that are not in a supported archive format used
	  for testing.

0.28	18/09/2008
	- added SQLite pragma 'auto_vacuum' to database creation.
	- created an _init function to generate db/nntp objects in one place
	- add --localonly option to help screen in bin/cpanstats-verify (v0.19)
	- fixed test script to correctly do cmp_ok :)

0.27	11/09/2008
	- added 'grade' option to search in bin/cpanstats-select (v0.08)
	- fixed help text in bin/cpanstats-update (v0.06)
	- fixed test bug in 12generate.t (I can't count!)

0.26	02/09/2008
	- abstracted all database functionality out of Generator.pm into
	  Database.pm, including database creation. All scripts updated too.
	- database links changed to reference CPAN Testers development site.
	- corrections to the database schema in POD
	- several POD updates across all files, including
	    bin/cpanstats	  (v0.04)
	    bin/cpanstats-delete  (v0.11)
	    bin/cpanstats-reparse (v0.07)
	    bin/cpanstats-select  (v0.07)
	    bin/cpanstats-update  (v0.05)
	    bin/cpanstats-verify  (v0.18)
	- fixed dependency in cpanstats
	- consistently provided help|h and version|V options to all scripts
	- fix to support old form of PAUSE uploads
	- abstracted reparse code into Generator.pm from cpanstats-reparse
	- more tests

0.25	28/08/2008
	- added POD about bugs and see also sections.
	- fixed POD in scripts.
	- added author to upload metadata, to save lookups elsewhere.
	- renamed several supporting scripts:
	    delstats    => bin/cpanstats-delete (v0.10)
	    selectstats => bin/cpanstats-select (v0.06)
	    upstats     => bin/cpanstats-update (v0.04)
	    readstats   => bin/cpanstats-verify (v0.17)
	- added further functionality to cpanstats-select.
	- added bin/cpanstats-reparse (v0.06) to distribution
	- added nostore option to not retain old articles in the articles
	  database.

0.24	19/08/2008
	- added bin/readstats (v0.16), bin/delstats (v0.09),
          bin/selectstats (v0.05) and bin/upstats (v0.03) from the non-CPAN
          cpanstats distribution.
        - updated all scripts to use an options hash rather than individual
          variables.
        - added new Database.pm module to handle all local DB interaction.
        - fixed test bug in t/12generate.t (thanks to Slaven Rezic).

0.23	18/08/2008
	- new maintainer: Barbie.
	- major overhaul to draw in CPAN Testers Statistics code.
	- added version to Article.pm.
	- added ability to decode QuotedPrintable and Base64 (Article.pm)
	- added functionality to parse the upload articles as well as the
	  report articles (Article.pm).
	- added more reliable subject parsing.
	- added more reliable Perl -V parsing.
	- testers.db now renamed to cpanstats.db.
	- news.db now renamed to articles.db.
	- field names changed: 
	    status => state, 
	    distribution => dist, 
	    archname => platform.
	- new fields added: postdate,tester.
	- fixed prerequisites in modules, Makefile.PL and META.yml.
	- bin/cpan_www_testers_generate replaced with bin/cpanstats.pl
	- added more test samples
	- added more tests

0.22	Thu Feb  3 15:27:41 GMT 2005
	- now includes perl + OS version in the testers.db
	- use SQLite 3.0
	- bin/cpan_www_testers_generate expanded to give more
	  flexibility for execution
	- added Makefile.PL (as per TODO)
	- added tests
	- added more POD
	- (thanks to Barbie)

0.21	Sun Mar  7 18:31:35 GMT 2004
	- the NNTP cache now uses SQLite instead of DB_File
	- rename Changes to CHANGES

0.20	Wed Aug 13 21:52:06 BST 2003
	- make sure to HTML escape everything
	- design update, now with CSS
	- Fix bug where reports are use A/AM/AMS/Crypt-TEA-1.22.tar.gz
	  instead of Crypt-TEA-1.22.tar.gz
	- added a TODO
	- sort using version, not Sort::Version
	- renamed to CPAN::WWW::Testers::Generator
	- included HISTORY section by Elaine Ashton
	- moved examples/generate.pl to an installed script,
	  bin/cpan_www_testers_generate

0.10	Thu Jul 31 23:09:25 BST 2003
	- initial release


lib/CPAN/WWW/Testers/Generator.pm
package CPAN::WWW::Testers::Generator;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.31';

q(Become a CPAN Tester - http://wiki.cpantesters.org/wiki/GettingStarted);

__END__

=head1 NAME

CPAN::WWW::Testers::Generator - The pre-2009 CPAN Testers data manager codebase

=head1 DESCRIPTION

This distribution used to generate the CPAN Testers database before 2009.

=head1 CPAN TESTERS

The CPAN Testers project was conceived back in May 1998 by Graham Barr and 
Chris Nandor as a way to provide multi-platform testing for modules uploaded
to CPAN. Today there are over 20 million test reports and more than 100 
testers regularly submitting reports each month, giving valuable feedback 
for users and authors alike.
  
The objective of the group is to test as many of the distributions 
available on CPAN as possible, on as many platforms as possible, with a 
variety of perl interpreters. The ultimate goal is to improve the 
portability of the distributions on CPAN, and provide good feedback to the 
authors.

This distribution was originally written by Leon Brocard to download and
summarize CPAN Testers data. However, since 2008, the codebase has been
redesigned by Barbie, firstly to use the cpanstats database, and secondly
to retrieve data from the Metabase API.

Following the complete redesign of the tools in 2009, this distribution 
and the tools used were no longer required.

For the current set of releases for the CPAN Testers family of websites,
please see:

  * CPAN-Testers-WWW-Reports
  * CPAN-Testers-WWW-Statistics
  * CPAN-Testers-WWW-Wiki
  * CPAN-Testers-WWW-Blog

ADVANCED WARNING - ADVANCED WARNING - ADVANCED WARNING - ADVANCED WARNING

This distribution will soon be deleted from CPAN. It is no longer
applicable for the building of the CPAN Testers websites, and no longer
maintained.

=head1 SEE ALSO

=over

=item * L<CPAN-Testers-WWW-Reports>     - F<http://www.cpantesters.org>

=item * L<CPAN-Testers-WWW-Statistics>  - F<http://stats.cpantesters.org>

=item * L<CPAN-Testers-WWW-Wiki>        - F<http://wiki.cpantesters.org>

=item * L<CPAN-Testers-WWW-Blog>        - F<http://blog.cpantesters.org>

=back

=head1 BECOME A TESTER

The objective of the CPAN Testers is to test as many of the distributions 
on CPAN as possible, on as many platforms as possible. The ultimate goal is
to improve the portability of the distributions on CPAN, and provide good 
feedback to the authors.

Whether you have a common platform or a very unusual one, you can help by 
testing modules you install and submitting reports. There are plenty of 
module authors who could use test reports and helpful feedback on their 
modules and distributions. 

If you'd like to get involved, please take a look at the CPAN Testers Wiki,
where you can learn how to install and configure one of the recommended 
smoke tools.

For further help and advice, please subscribe to the the CPAN Testers 
discussion mailing list.

CPAN Testers Wiki - F<http://wiki.cpantesters.org>
CPAN Testers Discuss mailing list
- F<http://lists.cpan.org/showlist.cgi?name=cpan-testers-discuss>

=head1 AUTHOR

  Original author:    Leon Brocard <acme@astray.com>   2002-2008
  Current maintainer: Barbie       <barbie@cpan.org>   2008-2012

=head1 LICENSE

This code is distributed under the same license as Perl.


lib/CPAN/WWW/Testers/Generator/Database.pm
package CPAN::WWW::Testers::Generator::Database;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.29';

#----------------------------------------------------------------------------

=head1 NAME

CPAN::WWW::Testers::Generator::Database - DB handling code.

=head1 SYNOPSIS

  my $dbi = CPAN::WWW::Testers::Generator::Database->new(database => $db);
  my @rows = $dbi->get_query($sql);
  $dbi->do_query($sql);

  my $iterator = $dbi->get_query_interator($sql);
  while(my $row = $iterator->()) {
    # do something
  }

=head1 DESCRIPTION

Database handling code for interacting with a local cpanstats database.

=cut

# -------------------------------------
# Library Modules

use DBI;
use File::Basename;
use File::Path;

# -------------------------------------
# Variables

use constant    DATABASE    => 'cpanstats.db';

# -------------------------------------
# Routines

=head1 INTERFACE

=head2 The Constructor

=over 4

=item new

=back

=cut

sub new {
    my ($class,%hash) = @_;
    return  unless($hash{database});

    my $self = {database => $hash{database}};
    bless $self, $class;

    $self->{AutoCommit} = $hash{AutoCommit} || 0;

    my $exists = -f $self->{database};

    mkpath(dirname($self->{database}))  unless($exists);

    $self->{dbh} = DBI->connect("DBI:SQLite:dbname=$self->{database}", "", "", {
        RaiseError => 1,
        AutoCommit => $self->{AutoCommit},
        sqlite_handle_binary_nulls => 1,
    });
    return  unless($self->{dbh});

    if(!$exists) {
        eval { $self->_dbh_create($self->{dbh},$self->{database}) };
        die "Failed to create database: $@"  if($@);
    }

    return $self;
}

sub DESTROY {
    my $self = shift;
    return      unless($self->{dbh});

    $self->{dbh}->commit    unless($self->{AutoCommit});
    $self->{dbh}->disconnect;
}

=head2 Methods

=over 4

=item do_commit

Force a commit if AutoCommit is off

=cut

sub do_commit {
    my $self = shift;
    $self->{dbh}->commit;
}

=item do_query

An SQL wrapper method to perform a non-returning request.

=cut

sub do_query {
    my ($self,$sql,@fields) = @_;
    my $sth;

    # prepare the sql statement for executing
    eval { $sth = $self->{dbh}->prepare($sql); };
    unless($sth) {
        die sprintf "ERROR: %s : %s\n", $self->{dbh}->errstr, $sql;
    }

    # execute the SQL using any values sent to the function
    # to be placed in the sql
    unless($sth->execute(@fields)) {
        die sprintf "ERROR: %s : %s : [%s]\n", $sth->errstr, $sql, join(',',@fields);
    }

    $sth->finish;
}

=item get_query

An SQL wrapper method to perform a returning request.

=cut

sub get_query {
    my ($self,$sql,@fields) = @_;
    my ($sth,@rows);

    eval { $sth = $self->{dbh}->prepare($sql); };
    unless($sth) {
        die sprintf "ERROR: %s : %s\n", $self->{dbh}->errstr, $sql;
    }

    unless($sth->execute(@fields)) {
        die sprintf "ERROR: %s : %s : [%s]\n", $sth->errstr, $sql, join(',',@fields);
    }

    while(my $row = $sth->fetchrow_arrayref) {
        push @rows, [@$row];
    }
    return @rows;
}

=item get_query_iterator

An SQL wrapper method to perform a returning request, via an iterator.

=cut

sub get_query_iterator {
    my ($self,$sql,@fields) = @_;
    my ($sth,@rows);

    eval { $sth = $self->{dbh}->prepare($sql); };
    unless($sth) {
        die sprintf "ERROR: %s : %s\n", $self->{dbh}->errstr, $sql;
    }

    unless($sth->execute(@fields)) {
        die sprintf "ERROR: %s : %s : [%s]\n", $sth->errstr, $sql, join(',',@fields);
    }

    return sub { return $sth->fetchrow_arrayref }
}

sub _dbh_create {
    my ($self,$dbh,$db) = @_;
    my @sql;

    if($db =~ /cpanstats.db$/) {
        push @sql,
            'PRAGMA auto_vacuum = 1',
            'CREATE TABLE cpanstats (
                          id            INTEGER PRIMARY KEY,
                          state         TEXT,
                          postdate      TEXT,
                          tester        TEXT,
                          dist          TEXT,
                          version       TEXT,
                          platform      TEXT,
                          perl          TEXT,
                          osname        TEXT,
                          osvers        TEXT,
                          date          TEXT)',

            'CREATE INDEX distverstate ON cpanstats (dist, version, state)',
            'CREATE INDEX ixperl ON cpanstats (perl)',
            'CREATE INDEX ixplat ON cpanstats (platform)',
            'CREATE INDEX ixdate ON cpanstats (postdate)';
    } else {
        push @sql,
            'PRAGMA auto_vacuum = 1',
            'CREATE TABLE articles (
                          id            INTEGER PRIMARY KEY,
                          article       TEXT)';
    }

    $dbh->do($_)    for(@sql);
}


__END__

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-WWW-Testers-Generator

=head1 SEE ALSO

L<CPAN::WWW::Testers>,
L<CPAN::Testers::WWW::Statistics>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  for Miss Barbell Productions <http://www.missbarbell.co.uk>.

=head1 COPYRIGHT AND LICENSE

  Copyright (C) 2008 Barbie for Miss Barbell Productions.

  This module is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut



lib/CPAN/WWW/Testers/Generator/Article.pm
package CPAN::WWW::Testers::Generator::Article;

use warnings;
use strict;
use vars qw($VERSION);

$VERSION = '0.29';

#----------------------------------------------------------------------------
# Library Modules

use MIME::Base64;
use MIME::QuotedPrint;
use CPAN::DistnameInfo;
use Email::Simple;

use base qw( Class::Accessor::Fast );

#----------------------------------------------------------------------------
# Variables

my %month = (
	Jan => 1, Feb => 2, Mar => 3, Apr => 4,  May => 5,  Jun => 6,
	Jul => 7, Aug => 8, Sep => 9, Oct => 10, Nov => 11, Dec => 12,
);

#----------------------------------------------------------------------------
# The Application Programming Interface

__PACKAGE__->mk_accessors(qw(   postdate date status from distribution version
                                perl osname osvers archname subject author));

sub new {
    my($class, $article) = @_;
    my $self = {};
    bless $self, $class;

    $article = decode_qp($article)	if($article =~ /=3D/);

    my $mail = Email::Simple->new($article);
    return unless $mail;
    return if $mail->header("In-Reply-To");

    my $from    = $mail->header("From");
    my $subject = $mail->header("Subject");
    return unless $subject;
    return if $subject =~ /::/; # it's supposed to be a distribution

    $self->{mail}    = $mail;
    $self->{from}    = $from;
    $self->{subject} = $subject;

    ($self->{postdate},$self->{date}) = _parse_date($mail);

    return $self;
}

sub _parse_date {
    my $mail = shift;
    my ($date1,$date2) = _extract_date($mail->header("Date"));
    my @received  = $mail->header("Received");

    for my $hdr (@received) {
        next    unless($hdr =~ /.*;\s+(.*)\s*$/);
        my ($dt1,$dt2) = _extract_date($1);
        if($dt2 > $date2 + 1200) {
            $date1 = $dt1;
            $date2 = $dt2;
        }
    }

#print STDERR "        ... X.[Date: ".($date||'')."]\n";
    return($date1,$date2);
}

my %regexes = (
    # with time
    1 => { re => qr/(?:\w+,)?\s+(\d+)\s+(\w+)\s+(\d+)\s+(\d+):(\d+)/,   f => [qw(day month year hour min)] },     # Wed, 13 September 2004 06:29
    2 => { re => qr/(\d+)\s+(\w+)\s+(\d+)\s+(\d+):(\d+)/,               f => [qw(day month year hour min)] },     # 13 September 2004 06:29
    3 => { re => qr/(\w+)?\s+(\d+),?\s+(\d+)\s+(\d+):(\d+)/,            f => [qw(month day year hour min)] },     # September 22, 1999 06:29

    # just the date
    4 => { re => qr/(?:\w+,)?\s+(\d+)\s+(\w+)\s+(\d+)/, f => [qw(day month year)] },  # Wed, 13 September 2004
    5 => { re => qr/(\d+)\s+(\w+)\s+(\d+)/,             f => [qw(day month year)] },  # 13 September 2004
    6 => { re => qr/(\w+)?\s+(\d+),?\s+(\d+)/,          f => [qw(month day year)] },  # September 22, 1999 06:29
);

sub _extract_date {
    my $date = shift;
    my (%fields,@fields,$index);

#print STDERR "#        ... 0.[Date: ".($date||'')."]\n";

    for my $inx (sort {$a <=> $b} keys %regexes) {
        (@fields) = ($date =~ $regexes{$inx}->{re});
        if(@fields) {
            $index = $inx;
            last;
        }
    }

    return('000000','000000000000') unless(@fields && $index);

    @fields{@{$regexes{$index}->{f}}} = @fields;

    $fields{month} = substr($fields{month},0,3);
    return('000000','000000000000') unless($month{$fields{month}} && $fields{year} > 1998);

    $fields{$_} ||= 0   for(qw(year month day hour min));

#print STDERR "#        ... 1.[$_][$fields{$_}]\n"   for(qw(year month day hour min));
    my $short = sprintf "%04d%02d",             $fields{year}, $month{$fields{month}};
    my $long  = sprintf "%04d%02d%02d%02d%02d", $fields{year}, $month{$fields{month}}, $fields{day}, $fields{hour}, $fields{min};

    return($short,$long);
}

sub parse_upload {
    my $self = shift;
    my $mail = $self->{mail};
    my $subject = $self->{subject};

    return 0	unless($subject =~ /CPAN Upload:\s+([-\w\/\.\+]+)/i);
    my $distvers = $1;

    # only record supported archives
    return 0    if($distvers !~ /\.(?:(?:tar\.|t)(?:gz|bz2)|zip)/);

    # CPAN::DistnameInfo doesn't support .tar.bz2 files ... yet
    $distvers =~ s/\.(?:tar\.|t)bz2//i;
    $distvers .= '.tar.gz' unless $distvers =~ /\.(?:(?:tar\.|t)gz|zip)/i;

    # CPAN::DistnameInfo doesn't support old form of uploads
    my @parts = split("/",$distvers);
    if(@parts == 2) {
        my ($first,$second,$rest) = split(//,$distvers,3);
        $distvers = "$first/$first$second/$first$second$rest";
    }

    my $d = CPAN::DistnameInfo->new($distvers);
    $self->distribution($d->dist);
    $self->version($d->version);
    $self->author($d->cpanid);

    return 1;
}

sub parse_report {
    my $self = shift;
    my $mail = $self->{mail};
    my $from = $self->{from};
    my $subject = $self->{subject};

    my ($status, $distversion, $platform) = split /\s+/, $subject;
    return 0  unless $status =~ /^(PASS|FAIL|UNKNOWN|NA)$/i;

    $platform ||= "";
    $platform =~ s/[\s&,<].*//;

    $distversion =~ s!/$!!;
    $distversion =~ s/\.tar.*/.tar.gz/;
    $distversion .= '.tar.gz' unless $distversion =~ /\.(tar|tgz|zip)/;

    my $d = CPAN::DistnameInfo->new($distversion);
    my ($dist, $version) = ($d->dist, $d->version);
    return 0 unless defined $dist;
    return 0 unless defined $version;

    my $encoding = $mail->header('Content-Transfer-Encoding');

    my $body = $mail->body;
    $body = decode_base64($body)  if($encoding && $encoding eq 'base64');

    my $perl = $self->_extract_perl_version(\$body);

    my ($osname)   = $body =~ /Summary of my perl5.*osname=([^\s\n,<\']+)/s;
    my ($osvers)   = $body =~ /Summary of my perl5.*osvers=([^\s\n,<\']+)/s;
    my ($archname) = $body =~ /Summary of my perl5.*archname=([^\s\n&,<\']+)/s;
    $archname =~ s/\n.*//	if($archname);

    $self->status($status);
    $self->distribution($dist);
    $self->version($version);
    $self->from($from || "");
    $self->perl($perl);
    $self->osname($osname || "");
    $self->osvers($osvers || "");

    unless($archname || $platform) {
  	    if($osname && $osvers)	{ $platform = "$osname-$osvers" }
	    elsif($osname)		    { $platform = $osname }
    }

    $self->archname($archname || $platform);

    return 1;
}

sub passed {
    my $self = shift;
    return $self->status eq 'PASS';
}

sub failed {
    my $self = shift;
    return $self->status eq 'FAIL';
}

# there are a few old test reports that omitted the perl version number.
# In these instances 0 is assumed. These reports are now so old, that
# worrying about them is not worth the effort.

sub _extract_perl_version {
    my ($self, $body) = @_;

    # Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
    # Summary of my perl5 (revision 5 version 10 subversion 0) configuration:
    my ($rev, $ver, $sub, $extra) =
        $$body =~ /Summary of my (?:perl\d+)? \((?:revision )?(\d+(?:\.\d+)?) (?:version|patchlevel) (\d+) subversion\s+(\d+) ?(.*?)\) configuration/s;

    unless(defined $rev) {
#       warn "Cannot parse perl version for article:\n$body";
        return 0;
    }

    my $perl = $rev + ($ver / 1000) + ($sub / 1000000);
    $rev = int($perl);
    $ver = int(($perl*1000)%1000);
    $sub = int(($perl*1000000)%1000);

    my $version = sprintf "%d.%d.%d", $rev, $ver, $sub;
    $version .= " $extra" if $extra;
    return $version;
#   return sprintf "%0.6f", $perl;	# an alternate format
}

1;

__END__

=head1 NAME

CPAN::WWW::Testers::Generator::Article - Parse a CPAN Testers article

=head1 DESCRIPTION

This is used by CPAN::WWW::Testers::Generator.

=head1 INTERFACE

=head2 The Constructor

=over 4

=item * new

The constructor. Pass in a reference to the article.

=back

=head2 Methods

=over 4

=item * parse_upload

Parses an upload article.

=item * parse_report

Parses a report article.

=item * passed

Whether the report was a PASS

=item * failed

Whether the report was a FAIL

=back

=head2 Accessors

All the following are accessors available through via the object, once an
article has been parsed as a report or upload announcement.

=over 4

=item * postdate

'YYYYMM' representation of the date article was posted.

=item * date

'YYYYMMDDhhmm' representation of the date article was posted.

=item * status

For reports this will be the grade, for uploads this will be 'CPAN'.

=item * from

Who posted the article.

=item * distribution

The distribution name.

=item * version

The distribution version.

=item * perl

The perl interpreter version used for testing.

=item * osname

Operating system name.

=item * osvers

Operating system version.

=item * archname

Operating system architecture name. This is usually based on the osname and
osvers, but they are not always the same.

=item * subject

Subject line of the original post.

=back

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties, that is not explained within the POD
documentation, please send bug reports and patches to the RT Queue (see below).

Fixes are dependant upon their severity and my availablity. Should a fix not
be forthcoming, please feel free to (politely) remind me.

RT Queue -
http://rt.cpan.org/Public/Dist/Display.html?Name=CPAN-WWW-Testers-Generator

=head1 SEE ALSO

L<CPAN::WWW::Testers>,
L<CPAN::Testers::WWW::Statistics>

F<http://www.cpantesters.org/>,
F<http://stats.cpantesters.org/>,
F<http://wiki.cpantesters.org/>

=head1 AUTHOR

  Original author:    Leon Brocard <acme@astray.com>   (C) 2002-2008
  Current maintainer: Barbie       <barbie@cpan.org>   (C) 2008

=head1 LICENSE

This code is distributed under the same license as Perl.
